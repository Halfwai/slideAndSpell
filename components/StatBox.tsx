import { View, Text, StyleSheet, Animated, Dimensions, BackHandler } from 'react-native';
import MyAppText from '@/components/MyAppText';
import { COLOURS } from '@/constants/colours';
import { useRef, useEffect } from 'react';
import { RFPercentage } from "react-native-responsive-fontsize";

import { useRouter } from 'expo-router';

interface StatBoxProps {
    stat: string;
    value: number;
    position: 'left' | 'right' | 'bottom';
}

export default function StatBox(props: StatBoxProps) {
    const router = useRouter();

    // if(props.position === 'left') {
    let startPosition = {x: 0, y: 0};
    if (props.position === 'left') {
        startPosition = {x: -Dimensions.get('window').width, y: 0};
    } else if (props.position === 'right') {
        startPosition = {x: Dimensions.get('window').width, y: 0};
    } else if (props.position === 'bottom') {
        startPosition = {x: 0, y: Dimensions.get('window').height};
    }
    let position = useRef(new Animated.ValueXY(startPosition)).current;

    useEffect(() => {
        Animated.timing(position, {
            toValue: {x: 0, y: 0},
            duration: 1000,
            useNativeDriver: true,
        }).start();
    }, []);

    useEffect(() => {}, [
        BackHandler.addEventListener('hardwareBackPress', () => {
            Animated.timing(position, {
                toValue: startPosition,
                duration: 500,
                useNativeDriver: true,
            }).start();
            return false; // or return false based on your requirement
        })
    ]);



    return (
        <Animated.View style={[styles.statBox, {transform: [{translateX: position.x}, {translateY: position.y}]}]}>
            <MyAppText style={styles.title}>{props.stat}</MyAppText>
            <MyAppText style={styles.value}>{props.value === null || props.value === undefined ? "-" : props.value}</MyAppText>
        </Animated.View> 
    );
}

const styles = StyleSheet.create({
    statBox: {
        backgroundColor: 'white',
        padding: 10,
        margin: 5,
        borderRadius: 10,
        width: 150,
        alignItems: 'center',
        borderColor: COLOURS.green,
        borderWidth: 1,
        height: "100%",
        justifyContent: "center",
    },
    title: {
        fontSize: RFPercentage(2),
        textAlign: 'center',
    },
    value: {
        fontSize: RFPercentage(4),
        textAlign: 'center',
    }
});