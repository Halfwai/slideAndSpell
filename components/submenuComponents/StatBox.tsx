import React, { useRef, useEffect } from 'react';
import { StyleSheet, Animated, Dimensions, BackHandler } from 'react-native';
import { RFPercentage } from "react-native-responsive-fontsize";

// Import COLOURS
import { COLOURS } from '@/constants/colours';

// Import components
import MyAppText from '@/components/common/MyAppText';

// Setup props
interface StatBoxProps {
    stat: string;
    value: number | string;
    position: 'left' | 'right' | 'bottom';
}

export default function StatBox(props: StatBoxProps) {
    // Set the starting position of the stat box
    let startPosition = { x: 0, y: 0 };
    if (props.position === 'left') {
        startPosition = { x: -Dimensions.get('window').width, y: 0 };
    } else if (props.position === 'right') {
        startPosition = { x: Dimensions.get('window').width, y: 0 };
    } else if (props.position === 'bottom') {
        startPosition = { x: 0, y: Dimensions.get('window').height };
    }
    let position = useRef(new Animated.ValueXY(startPosition)).current;

    // Slide in the stat box
    useEffect(() => {
        Animated.timing(position, {
            toValue: { x: 0, y: 0 },
            duration: 1000,
            useNativeDriver: true,
        }).start();
    }, []);

    // Slide out the stat box when the back button is pressed
    useEffect(() => { }, [
        BackHandler.addEventListener('hardwareBackPress', () => {
            Animated.timing(position, {
                toValue: startPosition,
                duration: 500,
                useNativeDriver: true,
            }).start();
            return false; // or return false based on your requirement
        })
    ]);

    return (
        <Animated.View style={[styles.statBox, { transform: [{ translateX: position.x }, { translateY: position.y }] }]}>
            <MyAppText style={styles.title}>{props.stat}</MyAppText>
            <MyAppText style={styles.value}>{props.value === null || props.value === undefined ? "-" : props.value}</MyAppText>
        </Animated.View>
    );
}

const styles = StyleSheet.create({
    statBox: {
        backgroundColor: 'white',
        padding: 10,
        margin: 5,
        borderRadius: 10,
        width: 150,
        alignItems: 'center',
        borderColor: COLOURS.green,
        borderWidth: 1,
        height: "100%",
        justifyContent: "center",
    },
    title: {
        fontSize: RFPercentage(2),
        textAlign: 'center',
    },
    value: {
        fontSize: RFPercentage(4),
        textAlign: 'center',
    }
});