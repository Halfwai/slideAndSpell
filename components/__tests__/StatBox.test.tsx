import { render } from '@testing-library/react-native';
import StatBox from '@/components/submenuComponents/StatBox';

it(`StatBox Component renders correctly`, () => {
    const { toJSON } = render(
        <StatBox
            stat='Test Stat'
            value={100}
            position='left'
        />
    );    
    expect(toJSON()).toMatchSnapshot();
});