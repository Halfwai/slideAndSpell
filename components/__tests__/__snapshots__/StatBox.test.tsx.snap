// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StatBox Component renders correctly 1`] = `
<View
  collapsable={false}
  style={
    {
      "alignItems": "center",
      "backgroundColor": "white",
      "borderColor": "#8BF0BC",
      "borderRadius": 10,
      "borderWidth": 1,
      "height": "100%",
      "justifyContent": "center",
      "margin": 5,
      "padding": 10,
      "transform": [
        {
          "translateX": -750,
        },
        {
          "translateY": 0,
        },
      ],
      "width": 150,
    }
  }
>
  <Text
    style={
      [
        {
          "fontFamily": "Quicksand",
        },
        {
          "fontSize": 27,
          "textAlign": "center",
        },
      ]
    }
  >
    Test Stat
  </Text>
  <Text
    style={
      [
        {
          "fontFamily": "Quicksand",
        },
        {
          "fontSize": 53,
          "textAlign": "center",
        },
      ]
    }
  >
    100
  </Text>
</View>
`;
